### PR TITLE
Add HA support for OVN DBs using active/passive.

### PR DESCRIPTION
**Pull request by numansiddique**
Previous commit [1] added HA support for ovnkube-masters.
This patch adds HA support for OVN ovsdb-servers in active/standby mode
so that the OVN db contents are replicated. The ovnkube-master which
wins the leader election configures the OVN ovsdb-servers running in that
node to active and the ovnkube-masters which lost the election configures
the OVN ovsdb-servers running in that node to standby so that they replicate
the DB contents from the active.

A couple of configuration options are added: Important ones are :-
 - manage-db-servers (boolean) - If set, manages the OVN ovsdb-servers in active/standby
                                 mode. Uses the unix sockets to communicate with
                                 the ovsdb-servers.

 - pod-ip - The IP of pod allocated by k8s. Required when manage-db-servers is enabled.

If manage-db-servers option is set, it is expected that
 - OVN ovsdb-servers and ovn-northd are running in each node where ovnkube-master
   is running, there by tightly coupling all these services in the master node.
 - ovnkube-master can communicate with the ovsdb-servers and ovn-northd using
   the unix control socket. Eg. /var/run/openvswitch/ovnnb_db.ctl
 - ovn-northd is configured to connect to the local ovsdb-servers using the unix
   sockets (and not using the endpoint - ovnkube-db).

If manage-db-servers option is not set,
 - it is expected that the HA of OVN ovsdb-servers if desired is managed separately.
 - OVN ovsdb-servers and ovn-northd can be decoupled from ovnkube-master.

When the ovnkube-master is started, it participates in the
election. The ovnkube-master which wins the election, does the following
   - If manage-db-servers option is set, then
      * Creates or updates the endpoint - ovnkube-db with its IP address.
      * Configures the OVN ovsdb-servers as active, so that the OVN ovsdb-servers
        running in that node allow write transactions.
      * Resumes the ovn-northd running in that node, so that it can process the
        OVN NB DB entries and convert to the logical flows.
   - Starts monitoring the pod events.

There is no change required in ovnkube's running on the nodes as it monitors for
any changes to the ovnkube-db endpoint and updates the ovn-remote option so that
ovn-controller can connect to the active ovsdb-server.

[1] - c9a4d33eeb51("Add HA support for ovnkube masters.")

Signed-off-by: Numan Siddique <numans@ovn.org>

### CODE REVIEW

**Comment by numansiddique** [comment]
@alexanderConstantinescu  FYI.

**Comment by girishmg** [comment]
@dcbw taking a look now

**Comment by numansiddique** [comment]
@girishmg Hi, does this look good ? :)

**Comment by numansiddique** [comment]
> @numansiddique , I think this version of the code could use an added argument defining if HA should be enabled or not. I am saying this because I have an on-going PR for raft here: https://github.com/openshift/cluster-network-operator/pull/376/files which would break if your PR is integrated "as is".
> 

I don't understand your comment.

You mean a flag to participate in the election or not ?

How this will break your other PR in cluster-network-operator ?

> I am working on a PR for this repo which would use the same mechanism: to be filed today/tomorrow. Moreover I think that if we decide to go for raft "all out" (here and in down-stream openshift-ovn), we might need to modify your previous PR with leader election too, as it will only activate master for the elected leader. In raft, I think - and correct me if I am wrong - all masters should be active. `ovnkube-node` should be able to talk to all them, and they should all be active as to be able to respond.

I am not really sure if ovnkube-master support active/active. That is something @dcbw @squeed  or @girishmg  can confirm.

I think the reason I added leader election is because, only the leader will talk to the ovn db servers and create the necessary logical switches/routers etc.

If suppose you start a 3 replicas of ovnkube-masters, then all 3 of them would write to the db servers.

**Comment by numansiddique** [comment]
> If suppose you start a 3 replicas of ovnkube-masters, then all 3 of them would write to the db servers.

I mean if there is no leader election.

**Comment by alexanderConstantinescu** [comment]
Excuse my previous comment. I got confused between OVN DB HA and ovnkube HA. 

Concerning the PR: I will need to continue tomorrow. But in general I think error handling is the most important thing to fix for the moment, as to not log the same message twice in a lot of places.

File: go-controller/pkg/ovn/ha_master.go [hunk 1]
```diff
@@ -3,47 +3,73 @@ package ovn
...
-		kubeClient:    kubeClient,
-		ovnController: ovnController,
-		nodeName:      nodeName,
-		isLeader:      false,
-		leaderElector: nil,
-		stopChan:      stopChan,
+		kubeClient:      kubeClient,
+		ovnController:   ovnController,
+		nodeName:        nodeName,
+		manageDBServers: manageDBServers,
```
Comment (girishmg) [inline go-controller/pkg/ovn/ha_master.go]: set `manageDBServers:` to `config.MasterHA.ManageDBServers` directly at L51

File: go-controller/pkg/ovn/ha_master.go [hunk 2]
```diff
@@ -3,47 +3,73 @@ package ovn
...
+	"k8s.io/client-go/tools/cache"
 	"k8s.io/client-go/tools/leaderelection"
 	"k8s.io/client-go/tools/leaderelection/resourcelock"
 )
 
 const (
+	invalidIPAddress = "0.0.0.1"
 	haLeaderLockName = "ovn-kubernetes-master"
+	ovnkubeDbEp      = "ovnkube-db"
+	haMasterLeader   = "k8s.ovn.org.ovnkube-master-leader"
```
Comment (girishmg) [inline go-controller/pkg/ovn/ha_master.go]: @dcbw @numansiddique this should be `k8s.ovn.org/ovnkube-master-leader`

Comment (numansiddique) [reply go-controller/pkg/ovn/ha_master.go]: Done

File: go-controller/pkg/ovn/ha_master.go (hunk 1 above)
Comment (numansiddique) [reply go-controller/pkg/ovn/ha_master.go]: Done

File: go-controller/pkg/util/kube.go [hunk 3]
```diff
@@ -65,3 +65,31 @@ func ServiceTypeHasClusterIP(service *kapi.Service) bool {
...
+			sbDBPort = ovnDB.Port
+		} else if ovnDB.Name == "north" {
+			nbDBPort = ovnDB.Port
+		}
+	}
+	for _, address := range ep.Subsets[0].Addresses {
+		masterIPList = append(masterIPList, address.IP)
+	}
+
+	return masterIPList, sbDBPort, nbDBPort, nil
```
[the remark this answers is not in the document]
Comment (numansiddique) [reply go-controller/pkg/util/kube.go]: Ok. I am not sure if I have to done anything in this PR ?

File: go-controller/pkg/cluster/node.go [hunk 4]
```diff
@@ -180,46 +180,24 @@ func (cluster *OvnClusterController) StartClusterNode(name string) error {
...
+		logrus.Errorf(err.Error())
+		return err
 	}
 
-	if err := config.UpdateOVNNodeAuth(masterIPList, southboundDBPort, northboundDBPort); err != nil {
+	err = config.UpdateOVNNodeAuth(masterIPList, strconv.Itoa(int(southboundDBPort)), strconv.Itoa(int(northboundDBPort)))
+	if err != nil {
+		logrus.Errorf(err.Error())
 		return err
 	}
```
[the remark this answers is not in the document]
Comment (numansiddique) [reply go-controller/pkg/cluster/node.go]: Done

File: go-controller/pkg/cluster/node.go [hunk 5]
```diff
@@ -180,46 +180,24 @@ func (cluster *OvnClusterController) StartClusterNode(name string) error {
...
-		}
-	}
-
-	for _, address := range ep.Subsets[0].Addresses {
-		masterIPList = append(masterIPList, address.IP)
+	masterIPList, southboundDBPort, northboundDBPort, err := util.ExtractDbRemotesFromEndpoint(ep)
+	if err != nil {
+		logrus.Errorf(err.Error())
+		return err
 	}
```
[the remark this answers is not in the document]
Comment (numansiddique) [reply go-controller/pkg/cluster/node.go]: Done

File: go-controller/cmd/ovnkube/ovnkube.go [hunk 6]
```diff
@@ -225,7 +225,7 @@ func runOvnKube(ctx *cli.Context) error {
 		ovn.RegisterMetrics()
 
 		// run the HA master controller to init the master
-		ovnHAController := ovn.NewHAMasterController(clientset, factory, master, stopChan)
+		ovnHAController := ovn.NewHAMasterController(clientset, factory, master, stopChan, config.MasterHA.ManageDBServers)
 		if err := ovnHAController.StartHAMasterController(); err != nil {
```
[the remark this answers is not in the document]
Comment (numansiddique) [reply go-controller/cmd/ovnkube/ovnkube.go]: Done

**Comment by numansiddique** [comment]
@girishmg Thanks for the reviews. I addressed them except for your comment in go-controller/pkg/util/kube.go +L94.
I am not sure if i have to do anything in this PR.

Thanks for the reviews.

**Comment by girishmg** [comment]
@numansiddique @dcbw LGTM

File: go-controller/pkg/cluster/node.go [hunk 7]
```diff
@@ -169,46 +169,24 @@ func (cluster *OvnClusterController) StartClusterNode(name string) error {
...
+		logrus.Errorf(err.Error())
 		return err
 	}
 
 	for _, auth := range []config.OvnAuthConfig{config.OvnNorth, config.OvnSouth} {
 		if err := auth.SetDBAuth(); err != nil {
 			return err
 		}
-		logrus.Infof("OVN databases reconfigured, masterIP %s, northbound-db %s, southbound-db %s", ep.Subsets[0].Addresses[0].IP, northboundDBPort, southboundDBPort)
+		logrus.Infof("OVN databases reconfigured, masterIP %s, northbound-db %d, southbound-db %d", ep.Subsets[0].Addresses[0].IP, northboundDBPort, southboundDBPort)
```
[the remark this answers is not in the document]
Comment (numansiddique) [reply go-controller/pkg/cluster/node.go]: As of this patch, only the first IP in the endpoint is used.
May be once your patch on raft is merged, I think its better to update then.
I will leave AS IS for now.

File: go-controller/pkg/cluster/node.go [hunk 8]
```diff
@@ -169,46 +169,24 @@ func (cluster *OvnClusterController) StartClusterNode(name string) error {
...
-		if ovnDB.Name == "north" {
-			northboundDBPort = strconv.Itoa(int(ovnDB.Port))
-		}
-	}
-
-	for _, address := range ep.Subsets[0].Addresses {
-		masterIPList = append(masterIPList, address.IP)
+	masterIPList, southboundDBPort, northboundDBPort, err := util.ExtractDbRemotesFromEndpoint(ep)
+	if err != nil {
+		logrus.Errorf(err.Error())
```
[the remark this answers is not in the document]
Comment (numansiddique) [reply go-controller/pkg/cluster/node.go]: I think its already addressed.

File: go-controller/pkg/cluster/node.go [hunk 9]
```diff
@@ -169,46 +169,24 @@ func (cluster *OvnClusterController) StartClusterNode(name string) error {
...
+		return err
 	}
 
-	if err := config.UpdateOVNNodeAuth(masterIPList, southboundDBPort, northboundDBPort); err != nil {
+	err = config.UpdateOVNNodeAuth(masterIPList, strconv.Itoa(int(southboundDBPort)), strconv.Itoa(int(northboundDBPort)))
+	if err != nil {
+		logrus.Errorf(err.Error())
```
[the remark this answers is not in the document]
Comment (numansiddique) [reply go-controller/pkg/cluster/node.go]: Same. Its already addressed. Probably you have seen a little outdated code.

File: go-controller/pkg/ovn/ha_master.go [hunk 10]
```diff
@@ -107,13 +153,58 @@ func (hacontroller *HAMasterController) StartHAMasterController() error {
...
+		// Step 4: Update the ovnkube-db endpoints with the new master Ip.
+
+		// Find the endpoint for the service
+		ep, err := hacontroller.ovnController.kube.GetEndpoint(config.Kubernetes.OVNConfigNamespace, ovnkubeDbEp)
+		if err != nil {
+			ep = nil
+		}
+		err = hacontroller.updateOvnDbEndpoints(ep, true)
+		if err != nil {
+			logrus.Errorf("%s Endpoint create/update failed", ovnkubeDbEp)
```
[the remark this answers is not in the document]
Comment (numansiddique) [reply go-controller/pkg/ovn/ha_master.go]: Done

File: go-controller/pkg/ovn/ha_master.go [hunk 11]
```diff
@@ -122,3 +213,320 @@ func (hacontroller *HAMasterController) ConfigureAsActive(masterNodeName string)
 
 	return hacontroller.ovnController.Run()
 }
+
+//updateOvnDbEndpoints Updates the ovnkube-db endpoints. Should be called
+// only if ovnkube-master is leader. This function will create the ovnkube-db endpoints
+// if it doesn't exist.
+func (hacontroller *HAMasterController) updateOvnDbEndpoints(ep *kapi.Endpoints, configureInvalidIP bool) error {
```
[the remark this answers is not in the document]
Comment (numansiddique) [reply go-controller/pkg/ovn/ha_master.go]: It was something like this earlier - https://github.com/numansiddique/ovn-kubernetes/blob/ha_master_v2_p4/go-controller/pkg/cluster/ha_master.go#L151
and @dcbw  suggested to do this way :).

**Comment by numansiddique** [comment]
Thank you all for the reviews.

**Pull request merged**